### PR TITLE
refactor: hoist BasicBlock op-indexed decorators to the MastForest (1/2)

### PR DESCRIPTION
**Pull request by huitseeker**
On top of #2259, #2274, #2291, #2301, which make `MastNode`s immutable after insertion in a `MastForest`.

# BasicBlockNode Op-Indexed Decorators

This PR adds a new way to store decorator index information in `BasicBlockNode` that lets both old and new systems work together. Think of it like having two types of storage - one for blocks that stand alone, and one for blocks that are part of a `MastForest`.

## What Changed

We added a `DecoratorStore` that has two options:
- `Owned`: Stores decorator index pointers directly in the block (for standalone blocks)
- `Linked`: Points to shared decorator index information (for blocks in a `MastForest`)

This means blocks inside a `MastForest` can now represent their decorator index information in a single sparse vector (with a few side indexes), which saves memory and decreases allocator pressure.

## Key Improvements

**Memory Efficiency**: When many blocks contain little or no decorator index information, we represent them as a few indices in CSR representation, rather than with a per-node COO representation.

**Systematic Representation**: The new representation follows the doctrine (adapted to our context) of compressed, sparse, row-major "matrix" representations, which is a standard way to store sparse values efficiently.

**Custom Serialization**: Blocks can still be saved & loaded correctly, even with the new storage system.

<details><summary>**How It Works**</summary>

When you create a block by itself, via a Builder or manually, it uses `Owned` storage:

```rust
// legacy, don't do this outside tests
let test_block = BasicBlockNode::new(operations, decorators)?;
// Do this instead
let block = BasicBlockNodeBuilder::new(operations, decorators).build()?;
```

When you add that block to a MastForest, it automatically switches to `Linked` storage which defers to the `MastForest`'s centralized storage:

```rust
let node_id = BasicBlockNodeBuilder::new(operations, decorators).add_to_forest(&mut forest)?;
```

No need to change your code - everything works the same way.

### Testing

Unit tests make sure:
- All existing code still works
- Decorators are saved and loaded correctly
- Memory is laid out properly between blocks
- Edge cases like empty blocks work fine
- Several decorator arrangements (like multiple decorators on the same operation) work correctly

### References
- Contributes to #1776 and #1497.
- [plan](https://hackmd.io/@huitseeker/HkkhyiUsge)

</details>

## Benefits & Next steps

This change makes the basic block decorators uses less memory & iterate faster in some cases.

The next step would be:
- to eventually remove the old `Owned` storage completely, but for now both systems work together,
- to remove the `before_enter` and `after_exit` decorators out of the `MastNode`s, in a similar way.

### CODE REVIEW

File: core/src/mast/decorator_storage.rs [hunk 1]
```diff
@@ -0,0 +1,945 @@
...
+
+impl Default for DecoratorIndexMapping {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Immutable view over all `(op_idx, DecoratorId)` pairs for a node.
+/// Uses the two-level CSR encoded by `node_indptr_for_op_idx` and `op_indptr_for_dec_idx`.
+pub struct DecoratedLinks<'a> {
```
Comment (plafer) [inline core/src/mast/decorator_storage.rs:431]: This docstring (and others) talk about operation indices, but then refer to the general "node" (rather "basic block", the only node which has operations). FWIU this decorator storage will grow to include all nodes - so how does this work for non-basic block nodes for which `op_idx` doesn't apply?

File: core/src/mast/node/basic_block_node/mod.rs [hunk 2]
```diff
@@ -987,10 +1266,39 @@ impl BasicBlockNodeBuilder {
...
+            }))
+            .map_err(|_| MastForestError::TooManyNodes)?;
+
+        // Add decorator info to the forest storage
+        Arc::make_mut(&mut forest.decorator_storage)
+            .add_decorator_info_for_node(node_id, decorators_info)
+            .map_err(MastForestError::DecoratorError)?;
+
+        // Convert the block to Linked variant (this is the core of Step 2)
+        if let MastNode::Block(block_node) = &mut forest.nodes[node_id] {
```
Comment (plafer) [inline core/src/mast/node/basic_block_node/mod.rs]: I would create a `MastNode::unwrap_basic_block_mut()` to be used here

File: core/src/mast/node/basic_block_node/mod.rs [hunk 3]
```diff
@@ -987,10 +1266,39 @@ impl BasicBlockNodeBuilder {
...
+            block_node.decorators = DecoratorStore::Linked {
+                id: node_id,
+                storage: Arc::clone(&forest.decorator_storage),
+            };
```
Comment (plafer) [inline core/src/mast/node/basic_block_node/mod.rs]: IIUC, using `Arc::make_mut()` followed by `Arc::clone()` means that for every `BasicBlock` (and every node in the subsequent PR?), `Arc::make_mut()` will cause a `DecoratorIndexMapping::clone()` (such that we'll end up with many copies in memory).

Fixing this might need to wait for a larger refactor - at least for the obvious fix to use a plain `&mut DecoratorIndexMapping` during construction, and only put it behind an `Arc` at the end of `MastForest` construction. But maybe you see another way around it?

File: core/src/mast/node/basic_block_node/mod.rs [hunk 4]
```diff
@@ -27,6 +28,65 @@ use op_batch::OpBatchAccumulator;
...
+    /// which have not yet been inserted into a MAST forest.
+    Owned(DecoratorList),
+    /// The decorators are stored in a MAST forest and can be accessed via
+    /// this block's ID. The `Arc` provides shared, read-only access to the
+    /// forest's decorator storage.
+    Linked {
+        id: MastNodeId,
+        storage: Arc<DecoratorIndexMapping>,
+    },
+}
```
Comment (plafer) [inline core/src/mast/node/basic_block_node/mod.rs]: Is this meant as a long term solution, or a temporary one while waiting for the next PR? I'm basically trying to understand the path towards our ultimate goal of #1580 (and https://github.com/0xMiden/miden-vm/issues/1821#issuecomment-2920621174).

Or to recap, with this new `DecoratorStore`, we end up still storing decorator-related data within each node. But as I recall (one of) our ultimate goal is to be able to serialize all the decorator-related data at the end of the serialized representation, so that to remove this decorator/debug info data, we just need to truncate the serialized representation. As I understood it, #1776 is a step towards this goal: we first fully "separate" the decorator data from node data, which makes achieving that desired serialized representation more or less trivial.

Are we still working in this direction with this set of changes?

File: core/src/mast/node/basic_block_node/mod.rs (hunk 3 above)
Comment (huitseeker) [reply core/src/mast/node/basic_block_node/mod.rs]: Ooh, boy this was 🤡 . Thanks for pointing this out, I've advanced the schedule on the switch of all decorator-accessing methods to require a `&MastForest` in my latest addition to the PR, which incidentally revealed a bug on an edge case of the DecroatorIndexMapping data structure.

Good news: there's no `Arc` involved any more.

File: core/src/mast/node/basic_block_node/mod.rs (hunk 4 above)
Comment (huitseeker) [reply core/src/mast/node/basic_block_node/mod.rs]: At a high level, because we discussed this offline: #1580 is not in scope of the present PR, which resolutely takes the stance of being backwards compatible (take the custom `impl` of `serde::{Serialize, Deserialize}  for MastForest` as a witness of that strategy).

We're no longer storing decorator-related data within each node, because once the `MastForest` is finalized, all blocks are `Linked`. At that point, decorator information is a simple integer tag (their `MastNodeId`), which is the minimal amount of information needed to link that individual node to their decorator metadata in the `MastForest`-level `decorators` (which is keyed by that `MastNodeId`). 

Once we do decide to tackle changes in serialization, we really only have to work with the `MastForest::decorators` field: this is a `DecoratorIndexMapping` which only contains trivially (de)serializable information (vectors of small integers). 

So I would say this PR is a meaningful step towards #1580. I would wait to deal with `before_enter`/`after_exit` decorators before changing the serialization in one pass.

File: core/src/mast/decorator_storage.rs (hunk 1 above)
Comment (huitseeker) [reply core/src/mast/decorator_storage.rs:431]: > FWIU this decorator storage will grow to include all nodes

We will definitely use the same notion of CSR in the future, but in a simplified and separate data structure.
That is, `DecoratorIndexMapping` is a two-level CSR presenting the following mapping:
```
MastNodeId -> op_idx slice,
op_idx -> decorator_id slice,
```
We will have a simplified data structure for `before_enter`/`after_exit` decorators that will work with much of same algorithmics, but only represent a one-level mapping:
```
MastNodeIf -> decorator_id slice
```

I _might_ look into abstracting over the structure involved and see if I can nest two instances of a 1-level CSR to represent the nested CSR that `DecoratorIndexMapping` fundamentally is. But I haven't gone down that road so far.

File: core/src/mast/merger/mod.rs
Comment (plafer) [inline core/src/mast/merger/mod.rs:120]: Are we properly updating the `MastForest.decorator_storage`? I basically want to make sure that we're taking into account that `DecoratorIndexMapping.decorator_ids` will now be stale, and somehow we have to make sure we properly map those decorator IDs too. Might already be taken care of, but want to make sure we don't forget about it.

File: core/src/mast/node/basic_block_node/mod.rs [hunk 5]
```diff
@@ -85,14 +89,29 @@ pub struct BasicBlockNode {
     /// Multiple batches are used for blocks consisting of more than 72 operations.
     op_batches: Vec<OpBatch>,
     digest: Word,
-    #[cfg_attr(feature = "serde", serde(default, skip_serializing_if = "Vec::is_empty"))]
-    decorators: DecoratorList,
-    #[cfg_attr(feature = "serde", serde(default, skip_serializing_if = "Vec::is_empty"))]
+    /// This replaces the old `decorators` field
```
Comment (plafer) [inline core/src/mast/node/basic_block_node/mod.rs]: I would remove this sentence, since it's no longer clear what the "old" decorators field was

File: core/src/mast/mod.rs [hunk 6]
```diff
@@ -71,6 +72,10 @@ pub struct MastForest {
     /// codes, so they are stored in order to provide a useful message to the user in case a error
     /// code is triggered.
     error_codes: BTreeMap<u64, Arc<str>>,
+
+    /// Provides efficient access to decorators per operation per node during execution and
+    /// debugging.
+    decorator_storage: DecoratorIndexMapping,
```
Comment (plafer) [inline core/src/mast/mod.rs:78]: How about merging this field with `decorators` into a new type? The type would basically encapsulate everything needed to manage decorators in one place.

Alternatively, moving the `decorators` field down here would at least put them visually together.

File: core/src/mast/node/basic_block_node/mod.rs (hunk 5 above)
Comment (huitseeker) [reply core/src/mast/node/basic_block_node/mod.rs]: Done!

File: core/src/mast/mod.rs (hunk 6 above)
Comment (huitseeker) [reply core/src/mast/mod.rs:78]: Your first remark is approached in #2323. The second is done in an additional commit here.

File: core/src/mast/merger/mod.rs
Comment (huitseeker) [reply core/src/mast/merger/mod.rs:120]: The `decorator_id_mappings` vector in `MastForestMerger` tracks all remappings, and these are used to remap decorator IDs in nodes via `build_with_remapped_children()`. `remove_nodes` and its caller clean up the old decorator info. I've added a dedicated test `mast_forest_merge_op_indexed_decorators_preservation` to this stack of PRs, lmk if you like it.

File: core/src/mast/decorator_storage.rs [hunk 7]
```diff
@@ -0,0 +1,1129 @@
...
+                while it.peek().is_some_and(|(i, _)| *i == op) {
+                    self.decorator_ids.push(it.next().unwrap().1);
+                }
+            }
+            // final sentinel for this node
+            self.op_indptr_for_dec_ids.push(self.decorator_ids.len());
+
+            // Push end pointer for this node (index of last op pointer)
+            let end_ops = self.op_indptr_for_dec_ids.len() - 1;
+            let _ = self.node_indptr_for_op_idx.push(end_ops);
```
Comment (plafer) [inline core/src/mast/decorator_storage.rs]: We should `expect()` (or handle error if appropriate) instead of silently ignoring the error. Applies to every other place where we use this pattern

File: core/src/mast/node/basic_block_node/mod.rs (hunk 2 above)
Comment (huitseeker) [reply core/src/mast/node/basic_block_node/mod.rs]: This has been refactored to stg w/o any cloning.

File: core/src/mast/decorator_storage.rs (hunk 7 above)
Comment (huitseeker) [reply core/src/mast/decorator_storage.rs]: good catch, of course!

File: core/src/mast/decorator_storage.rs [hunk 8]
```diff
@@ -0,0 +1,1135 @@
...
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(all(feature = "arbitrary", test), miden_test_serde_macros::serde_test)]
+pub struct DecoratorIndexMapping {
+    /// All the decorator IDs per operation per node, in a CSR relationship with
+    /// node_indptr_for_op_idx and op_indptr_for_dec_ids
+    decorator_ids: Vec<DecoratorId>,
+    /// For the node which operation indices are in op_indptr_for_dec_ids[node_start, node_end],
+    /// the indices of its i-th operation are at decorator_ids[op_indptr_for_dec_ids[node_start
+    /// + i]..op_indptr_for_dec_ids[node_start + i + 1]]
```
Comment (adr1anh) [inline core/src/mast/decorator_storage.rs]: I think this would render better using quotes and perhaps a separate line?

File: core/src/mast/decorator_storage.rs [hunk 9]
```diff
@@ -0,0 +1,1135 @@
...
+/// operation per node.
+///
+/// This structure provides efficient access to decorator IDs in a hierarchical manner:
+/// 1. First level: Node -> Operations (operations belong to nodes)
+/// 2. Second level: Operation -> Decorator IDs (decorator IDs belong to operations)
+///
+/// The data layout follows CSR format at both levels:
+/// - `decorator_ids`: Flat storage of all DecoratorId values
+/// - `op_indptr_for_dec_ids`: Pointer indices for operations within decorator_ids
+/// - `node_indptr_for_op_idx`: Pointer indices for nodes within op_indptr_for_dec_ids
```
Comment (Al-Kindi-0) [inline core/src/mast/decorator_storage.rs]: The description of the fields is very terse, I would try to write this in a more simple way, even if at the cost of being verbose (unless this is done somewhere else)

File: core/src/mast/decorator_storage.rs [hunk 10]
```diff
@@ -0,0 +1,1135 @@
...
+/// node_indptr_for_op_idx: [0, 2, 3]   // Node 0: [0,2], Node 1: [2,3]
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(all(feature = "arbitrary", test), miden_test_serde_macros::serde_test)]
+pub struct DecoratorIndexMapping {
+    /// All the decorator IDs per operation per node, in a CSR relationship with
+    /// node_indptr_for_op_idx and op_indptr_for_dec_ids
+    decorator_ids: Vec<DecoratorId>,
+    /// For the node which operation indices are in op_indptr_for_dec_ids[node_start, node_end],
```
Comment (Al-Kindi-0) [inline core/src/mast/decorator_storage.rs]: which -> who's ?

File: core/src/mast/decorator_storage.rs [hunk 11]
```diff
@@ -0,0 +1,1135 @@
...
+            }
+        }
+
+        for window in node_slice.windows(2) {
+            if window[0] > window[1] {
+                return Err(DecoratorIndexError::InternalStructure);
+            }
+        }
+
+        Ok(Self {
```
Comment (Al-Kindi-0) [inline core/src/mast/decorator_storage.rs:185]: Q: aren't there some restrictions on admissible inputs to this function? I am thinking not per component but between components

File: core/src/mast/decorator_storage.rs (hunk 8 above)
Comment (huitseeker) [reply core/src/mast/decorator_storage.rs]: addressed in 44308ee423c9f5bc38b6878ca4849bc2f23e3e7d

File: core/src/mast/decorator_storage.rs (hunk 9 above)
Comment (huitseeker) [reply core/src/mast/decorator_storage.rs]: addressed in 44308ee423c9f5bc38b6878ca4849bc2f23e3e7d

File: core/src/mast/decorator_storage.rs (hunk 11 above)
Comment (huitseeker) [reply core/src/mast/decorator_storage.rs:185]: Yes, good remark, I've made its purpose implicit by making it test-only.

File: core/src/mast/decorator_storage.rs [hunk 12]
```diff
@@ -0,0 +1,1174 @@
...
+///
+/// The data layout follows CSR format at both levels:
+///
+/// - `decorator_ids` contains all the DecoratorId values in a single flat array. These are the
+///   actual decorator identifiers that need to be accessed. We store them contiguously to minimize
+///   memory overhead and improve cache locality when iterating.
+///
+/// - `op_indptr_for_dec_ids` stores pointer indices that map operations to their position within
+///   the `decorator_ids` array. For each operation, it contains the start index where that
+///   operation's decorator IDs begin in the flat storage.
```
Comment (Al-Kindi-0) [inline core/src/mast/decorator_storage.rs:26]: nit: what about the final value in this array, which is a bit special, right?

File: core/src/mast/decorator_storage.rs [hunk 13]
```diff
@@ -0,0 +1,1174 @@
...
+///
+/// - `node_indptr_for_op_idx` stores pointer indices that map nodes to their position within the
+///   `op_indptr_for_dec_ids` array. For each node, it contains the start index where that node's
+///   operation pointers begin.
```
Comment (Al-Kindi-0) [inline core/src/mast/decorator_storage.rs:30]: same comment as above

File: core/src/mast/decorator_storage.rs [hunk 14]
```diff
@@ -0,0 +1,1174 @@
...
+/// Node 0, Op 0: [decorator_id_0, decorator_id_1]
+/// Node 0, Op 1: [decorator_id_2]
+/// Node 1, Op 0: [decorator_id_3, decorator_id_4, decorator_id_5]
+/// ```
+///
+/// This would be stored as:
+/// ```text
+/// decorator_ids:         [0, 1, 2, 3, 4, 5]
+/// op_indptr_for_dec_ids: [0, 2, 3, 6]  // Node 0: ops [0,2], [2,3]; Node 1: ops [3,6]
+/// node_indptr_for_op_idx: [0, 2, 3]   // Node 0: [0,2], Node 1: [2,3]
```
Comment (Al-Kindi-0) [inline core/src/mast/decorator_storage.rs:49]: the comment on the comment seem to need updating as well

File: core/src/mast/decorator_storage.rs [hunk 15]
```diff
@@ -0,0 +1,1174 @@
...
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(all(feature = "arbitrary", test), miden_test_serde_macros::serde_test)]
+pub struct DecoratorIndexMapping {
+    /// All the decorator IDs per operation per node, in a CSR relationship with
+    /// node_indptr_for_op_idx and op_indptr_for_dec_ids
+    decorator_ids: Vec<DecoratorId>,
+    /// For the node whose operation indices are in
+    /// `op_indptr_for_dec_ids[node_start..node_end]`,
+    /// the indices of its i-th operation are at:
```
Comment (Al-Kindi-0) [inline core/src/mast/decorator_storage.rs:60]: nit: I would add something like "where `node_end` is `next_node_start` ..." with the convention about the final node index

File: core/src/mast/decorator_storage.rs [hunk 16]
```diff
@@ -0,0 +1,1174 @@
...
+    /// ```text
+    /// decorator_ids[op_indptr_for_dec_ids[node_start + i]..
+    ///               op_indptr_for_dec_ids[node_start + i + 1]]
+    /// ```
+    op_indptr_for_dec_ids: Vec<usize>,
+    /// The decorated operation indices for the n-th node are at
+    /// `op_indptr_for_dec_ids[n..n+1]`
```
Comment (Al-Kindi-0) [inline core/src/mast/decorator_storage.rs:67]: does this need updating as well?

**Review by Al-Kindi-0 — COMMENTED** [review]
Looks great

File: core/src/mast/decorator_storage.rs (hunk 12 above)
Comment (huitseeker) [reply core/src/mast/decorator_storage.rs:26]: This is detailed on the comment on the field `op_indptr_for_dec_ids` you're looking at the struct-level comment.

File: core/src/mast/decorator_storage.rs (hunk 13 above)
Comment (huitseeker) [reply core/src/mast/decorator_storage.rs:30]: Same answer.

File: core/src/mast/decorator_storage.rs (hunk 16 above)
Comment (huitseeker) [reply core/src/mast/decorator_storage.rs:67]: Yes!

File: core/src/mast/decorator_storage.rs (hunk 14 above)
Comment (huitseeker) [reply core/src/mast/decorator_storage.rs:49]: Will add unit test elucidating this in my next PR.
